[ADD] magentoerpconnect_pricelist_currency

hugosantosred opened the pull request:
- This module is intended to import Magento Orders in the same currency that they were created in magento by using price list mappings.

hugosantosred committed:
[ADD] magentoerpconnect_pricelist_currency

- This module is intended to import Magento Orders in the same currency that they were created in magento by using price list mappings.

guewen commented on magentoerpconnect_pricelist_currency/model/sale.py:
```diff
@@ -0,0 +1,58 @@
+    magento_currency_id = fields.Many2one('res.currency', 'Magento Currency')
+
+
+@magento(replacing=SaleOrderImportMapper)
```
When you use `@magento(replacing=SaleOrderImportMapper)`, it introduces a compatibility issue.
If another addon does the same thing, only one of them will win and effectively replace `SaleOrderImportMapper`.

As there is extensibility between backend versions (inheritance through parent backends), it is hard to have modularity for one backend version.

Usually when I encounter a similar issue, my process is the following:
1. Can this feature be integrated in the core module? If yes, it simplifies.
2. If not, I introduce a sort of hook in the core, example: https://github.com/OCA/connector-magento/blob/8.0/magentoerpconnect/sale.py#L430-L435, https://github.com/OCA/connector-magento/blob/8.0/magentoerpconnect/sale.py#L733-L735, in your case it could be SalePricelistImportMapper and SalePricelistImporter. Then I replace only the new ConnectorUnit, reducing the risks of conflicts (still possible but anyway if 2 addons want to change the behavior of the import of pricelists, they are maybe not compatible by design).

guewen replied:
Perhaps you could `replace` https://github.com/OCA/connector-magento/blob/8.0/magentoerpconnect/sale.py#L942-L948, but then your module would be incompatible with `magentoerpconnect_pricing` ( see https://github.com/OCA/connector-magento/blob/8.0/magentoerpconnect_pricing/sale.py#L27-L41)

guewen commented on magentoerpconnect_pricelist_currency/view/product_pricelist_view.xml:
```diff
@@ -0,0 +1,66 @@
+        <record id="product_pricelist_mapping_tree_view" model="ir.ui.view">
+            <field name="name">product.pricelist.mapping.tree</field>
+            <field name="model">product.pricelist.mapping</field>
+            <field name="type">tree</field>
```
`<field name="type">tree</field>` is deprecated

guewen commented on magentoerpconnect_pricelist_currency/view/product_pricelist_view.xml:
```diff
@@ -0,0 +1,66 @@
+        <record id="product_pricelist_mapping_form_view" model="ir.ui.view">
+            <field name="name">product.pricelist.mapping.form</field>
+            <field name="model">product.pricelist.mapping</field>
+            <field name="type">form</field>
```
`<field name="type">form</field>` is deprecated

guewen commented on magentoerpconnect_pricelist_currency/model/product_pricelist.py:
```diff
@@ -0,0 +1,48 @@
+    _inherit = 'product.pricelist'
+
+    mapping_id = fields.Many2one('product.pricelist.mapping',
+                                 'Pricelist Mapping', select=True)
```
s/select/index/

guewen commented on magentoerpconnect_pricelist_currency/model/sale.py:
```diff
@@ -0,0 +1,58 @@
+        if binding.pricelist_id.currency_id != binding.magento_currency_id:
+            pricelist = binding.pricelist_id.get_pricelist_for_currency(
+                binding.magento_currency_id.id)
+            binding.write({'pricelist_id': pricelist.id})
```
Is there a reason for that to be in `_after_import` and not in the Mapper?

guewen commented:
Hi @hugosantosred 

Your addon is interesting and very complete (tests! :smile:)
I just have a concern with the usage of `replacing` which I commented in the diff.

Thanks for your PR

guewen commented:
(Travis build is red due to an issue with the docs build I'm currently investigating https://github.com/OCA/maintainer-quality-tools/pull/288#issuecomment-153738843)

hugosantosred commented:
@guewen Thanks for the review! I will made the changes that you comment as soon as I can.

hugosantosred committed:
Deleted deprecated <field name="type"></field> in views

hugosantosred committed:
magentoerpconnect_pricelist_currency: Replace select for index in field

hugosantosred committed:
Merge branch '8.0' of https://github.com/OCA/connector-magento into 8.0-pricelist-currency

hugosantosred committed:
Add Hooks for currencyImporter and PricelistCurrencyAssign

Two hooks has been added in magento core:
  - SaleOrderCurrencyImportMapper: Used to import magento currency in Sales. Does Nothing in Base
  - SaleOrderPricelistCurrencyAssign: A ConnectorUnit used in _after_import. Does nothing in Base

hugosantosred committed:
Change @magento replacing Classes

Now replaces new hooks introduced in base module for compatibility with another magento addons

hugosantosred commented:
@guewen Changes Done! I've introduced hooks in magentoerpconnect core for currency mapping and another in _after_import for pricelist selection. 

I have to do pricelist selection, based on currency, in _after_import method because I need to know the final selected currency in Odoo Sale (Can be different than base sale price list if customer has one selected).

guewen commented on magentoerpconnect_pricelist_currency/model/product_pricelist.py:
```diff
@@ -0,0 +1,48 @@
+        currency_pricelist = self.search([
+            ('currency_id', '=', currency_id),
+            ('mapping_id', '=', self.mapping_id.id)
+        ])
```
just spot one thing here: you could add `limit=1`, it also simplifies the next line to:

``` python
return currency_pricelist or self
```

guewen commented:
Thanks! I just added one last comment, apart that it sounds good now

hugosantosred committed:
Change search in price list selection based on currency

This search only returns one item in order to simplify return of this method

guewen commented:
:+1:
